[dbuf] Distributed Buffer

danking opened the pull request:
I need to write the CI integration, but this is my proposed distributed buffer service.

It's not resilient to failure at all.

`python3 -m dbuf 8`

Will create an 8-process dbuf. You can start follower cores on another machine with `python3 -m dbuf 8 --leader http://LEADER:LEADER_PORT`.

For a sense of the performance, the following has 10 co-routines each sending 10k messages of 40k bytes each using a buffer size of 5MB (so each co-routine holds about 5MB in memory before flushing):
```
# python3 scale_test.py 10 5 40000 10000
create
write aggregate-throughput:  0.333 GiB/s
read aggregate-throughput:  0.213 GiB/s
```
This is on my laptop over loopback with `python3 -m dbuf 4`. Note that we send 4 GB (10 * 10k * 40k bytes). Each core will buffer 512 MiB, so each server core will flush to disk twice (the scale test explicitly equally distributes the load, so each server core gets 1 GB).

I've got a Scala client as well which I'll add in another PR.

---

Update: same as above benchmark but I had to reduce the maximum data sent in one HTTP request to 1MiB:

```
write aggregate-throughput:  0.194 GiB/s
read aggregate-throughput:  0.135 GiB/s
```

I can no longer run the test on my laptop due to all the changes I made to dbuf to make it run in k8s. I don't know how much of the reduced throughput is due to the HTTP window size. I'll increase all the NGINX max request sizes at some point and retest.

danking commented:
@cseed ready for a look.

danking commented:
@cseed bump, this is holding up getting the shuffler in

danking commented:
The failure is due to the new memory requirements. Apparently the python script uses a lot more memory than I thought. Trying to find the magic number now. The tests were passing before the memory limits.

danking commented:
latest test yields:
```
write aggregate-throughput:  0.263 GiB/s
read aggregate-throughput:  0.083 GiB/s
```

cseed commented on dbuf/Dockerfile:
```diff
@@ -0,0 +1,8 @@
+RUN python3 -m pip install ./dbuf
+
+ENTRYPOINT ["python3", "-m", "dbuf"]
+CMD []
```
This seems unnecessary.

danking replied:
Prior images may have specified command arguments. It seems prudent to override them.

cseed commented on dbuf/dbuf/aiofiles.py:
```diff
@@ -0,0 +1,32 @@
+import hailtop.utils as utils
+
+
+class AIOFiles:
```
I'm kind of shocked this is needed.  Did you consider https://github.com/Tinche/aiofiles?

cseed replied:
This should probably go in hailtop.utils.

danking replied:
There has not been a recent release of that project. My memory is vague at this point but it wasn't working, and it's trivial to roll my own.

cseed commented on dbuf/dbuf/client.py:
```diff
@@ -0,0 +1,171 @@
+            'POST',
+            f'{server_url}/s/{self.id}/get',
+            json=key)
+        assert resp.status == 200
```
This are unnecesssary since you set `raise_for_status=True`.

cseed commented on dbuf/tests/Dockerfile:
```diff
@@ -0,0 +1,10 @@
+WORKDIR /home/hail
+RUN python3 -m pip install -U numpy
+
+ENTRYPOINT []
```
Unnecessary?

danking replied:
Again, memory vague at this point, but batch does not handle entry points correctly so it was easier to just kill the entry point inherited from batch so that I can specify the command to invoke in build.yaml.

cseed commented on dbuf/dbuf/dbuf.py:
```diff
@@ -0,0 +1,287 @@
+import aiohttp as ah
```
Everywhere else we just import as aiohttp.

cseed commented on dbuf/dbuf/retry_forever.py:
```diff
@@ -0,0 +1,19 @@
+from .logging import log
+
+
+async def retry_forever(f, msg=None):
```
Can we put this in utils with retry transient errors, etc?

cseed commented on dbuf/dbuf/retry_forever.py:
```diff
@@ -0,0 +1,19 @@
+            break
+        except Exception as exc:
+            if msg:
+                log.warning(msg(exc))
```
This won't include the stack trace.  I'd prefer `log.exception(msg if msg else 'caught exception')` or something.

danking replied:
I figure retry_forever shouldn't be an error level message though, and I'm pretty sure log exception is. I'll write something that grabs the stack trace and prints a warn level message.

cseed replied:
This would be awesome.  I think we use log.exception in a bunch of places for the exception details (good) that should just be warn or info.

cseed commented on dbuf/dbuf/retry_forever.py:
```diff
@@ -0,0 +1,19 @@
+        i = 1.1 ** i
+        await asyncio.sleep(0.100 * i)
+        # max 44.5s
+        if i < 64:
```
So I think these are bad defaults for this method.  They came from the batch client that was trying to minimize the delay introduced by polling I think.  I'd prefer increasing the delay by 2 and adding in some randomness (I've been using random.uninform(0.7, 1.3)) elsewhere.

danking replied:
OK, I'm using sleep_and_backoff

cseed commented on dbuf/setup.cfg:
```diff
@@ -0,0 +1,2 @@
+[flake8]
+max-line-length=120
```
Please use the global setup.cfg that disables this error.

cseed commented on dbuf/setup.py:
```diff
@@ -0,0 +1,15 @@
+    packages = find_packages(),
+    include_package_data=True,
+    install_requires=[
+        'aiohttp>=3.6,<3.7'
```
We don't normally specify dependencies in the per-module setup, but inherit them from docker requirements.txt.

cseed commented on dbuf/dbuf/dbuf.py:
```diff
@@ -0,0 +1,287 @@
+
+parser = argparse.ArgumentParser(description='distributed buffer')
+parser.add_argument('--name', type=str, help='my name in k8s, e.g. dbuf-0.dbuf', default='localhost')
+parser.add_argument('--leader', type=str, help='directory in which to store data', required=False)
```
I think the help string is wrong here.

cseed commented on dbuf/dbuf/dbuf.py:
```diff
@@ -0,0 +1,287 @@
+        self.shuffle_create_lock = asyncio.Lock()
+        self.app.add_routes([
+            web.post('/s', self.create),
+            web.post('/s/{session}', self.post),
```
Calling something post seems a bit too generic.  post_session?

danking replied:
I changed the name of `post_worker` to `register_worker`. Changing `post` to `post_session` creates asymmetry with `get`. Changing both to `get_session` seems unnecessarily verbose given the context.

I could change this one to `put` and the other to `get`, but there's a verb called PUT so that seems confusing. I could make this `set` instead of `post`, but that seems to imply that I'm modifying something with identity, which I am not.

I really feel like post and get are the right words. Post some data, get some data.

This is not the endpoint that creates a new session, that's `create` / `POST /s`. This endpoint posts data *into* an extant session, `post` / `POST /s/session`

cseed commented on dbuf/dbuf/client.py:
```diff
@@ -0,0 +1,171 @@
+            'POST',
+            f'{self.root_url}/s')
+        assert resp.status == 200
+        self.id = int(await resp.text())
```
Prefer this to be json.

danking replied:
This seems like more code for the same effect and unnecessarily invokes a json parser?
```python
id = await resp.json()
assert isinstance(id, int)
self.id = id
```

cseed commented on dbuf/dbuf/client.py:
```diff
@@ -0,0 +1,171 @@
+            off += n2
+        return result
+
+    async def getmany(self, keys, retry_delay=1):
```
`retry_delay` unused.

cseed commented on dbuf/dbuf/client.py:
```diff
@@ -0,0 +1,171 @@
+                data = await resp.read()
+                for v, j in zip(self._decode(data), (x[1] for x in batch)):
+                    results[j] = v
+        await asyncio.gather(*[get_from_server(server, keys)
```
Unbounded gathers are almost certainly not what you want.  hailtop.utils has a bounded_gather.  We should always be using that unless we know gather is suitably bounded.

danking replied:
How do you define "suitable bounded"?

cseed commented on dbuf/dbuf/dbuf.py:
```diff
@@ -0,0 +1,287 @@
+            struct.pack_into('i', out, off, n)
+            off += 4
+            await self.read(key, memoryview(out)[off:(off+n)])
+        await asyncio.gather(*[read_into(key, off) for key, off in zip(keys, offs)])
```
Unbounded gather.

cseed commented on dbuf/dbuf/dbuf.py:
```diff
@@ -0,0 +1,287 @@
+                    assert await text == f'{session_id}', f'{text}, {session_id}'
+                    log.info(f'successfully created shuffle on {worker} at {worker_url}')
+            log.info(f'getting shuffle lock for {session_id}')
+            async with self.shuffle_create_lock:
```
What's this lock about?  I'm skeptical.

danking replied:
It makes session creation sequential. I wanted to make sure if worker i says it created the session that it had created the session I'm currently trying to create. Without it, sessions race to create themselves, some returning before the eventual state of all workers having all sessions.

I could fix by having workers create a session if someone POSTs data in a session of which they're unaware. That also makes growing the cluster (not currently supported) simpler.

danking replied:
The whole distributed part of distributed buffer is, as you can tell, a big hack.

danking commented:
@johnc1231 Cotton is on vacation and Jackie is rather busy with batch. If you can review cotton's comments and convince yourself that I've addressed them appropriately and make any further comments of your own, I'd like to get this merged.

@cseed when you return, we can continue to address in subsequent PRs any inadequately addressed issues.

danking commented:
closing until it works
